cclient: fixed-sized trits uses flex_trit_t.

oopsmonk opened the pull request:
This PR fixed #410 

# Test Plan:
CI must pass

Commit:
cclient: fixed-sized trits for get_trytes.

th0br0 commented on ciri/api/api.c:
```diff
@@ -99,32 +99,39 @@ retcode_t iota_api_get_trytes(iota_api_t const *const api,
+    tmp_trits = trit_array_new_from_flex(iter->hash, NUM_TRITS_HASH,
+                                         FLEX_TRIT_SIZE_243);
+    if (!tmp_trits) {
+      return RC_API_NULL_POINTER;
```
this should return an OUT_OF_MEMORY error?

th0br0 commented on common/trinary/trit_array.c:
```diff
@@ -113,6 +113,29 @@ trit_array_p trit_array_new_from_trytes(tryte_t const *const trytes) {
+trit_array_p trit_array_new_from_flex(flex_trit_t const *const trits,
+                                      size_t num_trits, size_t num_bytes) {
+  trit_array_p trit_array;
+  trit_array = (trit_array_p)malloc(sizeof(struct _trit_array));
```
use `calloc`

th0br0 commented on common/trinary/trit_array.c:
```diff
@@ -113,6 +113,29 @@ trit_array_p trit_array_new_from_trytes(tryte_t const *const trytes) {
+    // errno = IOTA_OUT_OF_MEMORY
+    return NULL;
+  }
+  memset(trit_array, 0, sizeof(struct _trit_array));
```
not needed with `calloc`

th0br0 commented on ciri/api/api.c:
```diff
@@ -99,32 +99,39 @@ retcode_t iota_api_get_trytes(iota_api_t const *const api,
+  CDL_FOREACH(req->hashes, iter) {
     hash_pack_reset(&pack);
+    // flex_trit to trit_array_p for iota_tangle_transaction_load
+    tmp_trits = trit_array_new_from_flex(iter->hash, NUM_TRITS_HASH,
```
we should use a hash243_queue here and not a trit_array?

oopsmonk replied:
I should put a FIXME here. 
For fixing it, @thibault-martinez could you provide something like `iota_tangle_transaction_load` accept `flex_trits` or `hash243_queue`?

thibault-martinez replied:
I agree that it would be handy and I have thought about it.
The issue is that by accepting `flex_trit_t *` we lose the size attribute hence genericity. 
In the meantime, I really think there's no need for dynamic allocation here.

oopsmonk replied:
I thought `iota_tangle_transaction_load` will take `flex_trit_t`, so `trit_array_new_from_flex` is a temporary implementation which keeps code clean and can be removed easily. 
If `flex_trit_t` is not a plan for `iota_tangle_transaction_load`, I should change code here.

thibault-martinez commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -344,6 +344,32 @@ static retcode_t hash81_queue_to_json_array(hash81_queue_t queue,
+    cJSON_ArrayForEach(current_obj, json_item) {
+      if (current_obj->valuestring != NULL) {
+        trit_array_p tmp_ref =
+            trit_array_new_from_trytes((tryte_t*)current_obj->valuestring);
```
Can't you just use `flex_trits_from_trytes` instead of `trit_array_new_from_trytes ` ?

thibault-martinez commented on cclient/serialization/tests/test_json.c:
```diff
@@ -352,13 +352,16 @@ void test_serialize_get_trytes(void) {
   get_trytes_req_t* req = get_trytes_req_new();
-  get_trytes_req_add_hash(req, TEST_81_TRYRES_1);
+  trit_array_p tmp_hash =
+      trit_array_new_from_trytes((tryte_t*)TEST_81_TRYRES_1);
```
Dito

thibault-martinez commented on ciri/api/api.c:
```diff
@@ -99,32 +99,39 @@ retcode_t iota_api_get_trytes(iota_api_t const *const api,
+  CDL_FOREACH(req->hashes, iter) {
     hash_pack_reset(&pack);
+    // flex_trit to trit_array_p for iota_tangle_transaction_load
+    tmp_trits = trit_array_new_from_flex(iter->hash, NUM_TRITS_HASH,
```
Use a static alocated `trit_array_t` and just set `.trits = iter->hash`, i.e. no dynamic allocation nor `memcpy` needed

thibault-martinez replied:
See example here: https://github.com/iotaledger/entangled/blob/develop/ciri/api/api.c#L176

thibault-martinez commented on ciri/api/tests/test_get_trytes.c:
```diff
@@ -49,18 +49,20 @@ void test_get_trytes_not_found(void) {
   api.conf.max_get_trytes = 100;
 
-  get_trytes_req_add_hash(req, NULL_HASH);
+  trit_array_p null_hash = trit_array_new_from_trytes(NULL_HASH);
```
No need for `trit_array_p`

thibault-martinez commented on ciri/api/tests/test_get_trytes.c:
```diff
@@ -83,10 +85,14 @@ void test_get_trytes_max(void) {
-  get_trytes_req_add_hash(req, TX_1_OF_4_HASH);
-  get_trytes_req_add_hash(req, TX_2_OF_4_HASH);
+  trit_array_p tx1_hash = trit_array_new_from_trytes(TX_1_OF_4_HASH);
+  trit_array_p tx2_hash = trit_array_new_from_trytes(TX_2_OF_4_HASH);
```
Dito

thibault-martinez commented on common/trinary/trit_array.c:
```diff
@@ -113,6 +113,29 @@ trit_array_p trit_array_new_from_trytes(tryte_t const *const trytes) {
   return trit_array;
 }
 
+trit_array_p trit_array_new_from_flex(flex_trit_t const *const trits,
```
Not sure this function is really needed following my comments

oopsmonk replied:
No needed, if `iota_tangle_transaction_load` accept flex_trit.

thibault-martinez replied:
Understood, but what I meant was that we could achieve the exact same thing with no dynamic allocation and no memcpy.

Commit:
ciri: removed trit_array_new_from_flex.

cclient: using flex_trits_from_trytes instead of trit_array_new_from_trytes.

Commit:
ciri: using flex_trits_from_trytes.

thibault-martinez commented on ciri/api/tests/test_get_trytes.c:
```diff
@@ -46,11 +46,11 @@ void test_get_trytes_not_found(void) {
   get_trytes_req_t *req = get_trytes_req_new();
   get_trytes_res_t *res = get_trytes_res_new();
   tryte_t tx_trytes[NUM_TRYTES_SERIALIZED_TRANSACTION];
+  flex_trit_t null_hash[FLEX_TRIT_SIZE_243] = {FLEX_TRIT_NULL_VALUE};
```
This is not initializing all the values to `FLEX_TRIT_NULL_VALUE`, only the first one then zeros all the others.

oopsmonk replied:
oops, thanks.

Commit:
ciri: fixing array init issue.

Commit:
cclient: fixed-sized trits for attach_to_tangle.

Commit:
cclient: hash_queue in broadcast_transactions.

Commit:
cclient: using hash_queue in check_consistency.

Commit:
ciri: fixing test_broadcast_transactions failed caused by request API change.

Commit:
cclient: req/res using flex_trit_t

- using flex_trit_t for get_node_info, get_transactions_to_approve
- using hash containers for get_inclusion_state, store_transactions, get_tips
- flex_hash is deprecated in json serializer, will be removed from source code later.
- fixing ciri test failed which is caused by changing cclient api.

thibault-martinez commented on cclient/request/attach_to_tangle.c:
```diff
@@ -12,39 +12,53 @@ attach_to_tangle_req_t* attach_to_tangle_req_new() {
+                                         tryte_t const* const trunk) {
+  size_t len = flex_trits_from_trytes(req->trunk, NUM_TRITS_HASH, trunk,
+                                      NUM_TRYTES_HASH, NUM_TRYTES_HASH);
+  return len ? RC_OK : RC_CCLIENT_FLEX_TRITS;
```
You can go further and check that `len  == NUM_TRYTES_HASH `

oopsmonk replied:
`flex_trits_from_trytes` returns `0` or a specific tryte number(depended on `num_trytes`) only, and `0` means error.

thibault-martinez replied:
My mistake sorry, I thought it could return a number different than `num_trytes`.

thibault-martinez commented on cclient/request/attach_to_tangle.c:
```diff
@@ -12,39 +12,53 @@ attach_to_tangle_req_t* attach_to_tangle_req_new() {
+                                          tryte_t const* const branch) {
+  size_t len = flex_trits_from_trytes(req->branch, NUM_TRITS_HASH, branch,
+                                      NUM_TRYTES_HASH, NUM_TRYTES_HASH);
+  return len ? RC_OK : RC_CCLIENT_FLEX_TRITS;
```
You can go further and check that `len == NUM_TRYTES_HASH`

thibault-martinez commented on cclient/request/attach_to_tangle.c:
```diff
@@ -12,39 +12,53 @@ attach_to_tangle_req_t* attach_to_tangle_req_new() {
+  if (len) {
+    hash8019_queue_push(&req->trytes, trits);
+  }
+  return len ? RC_OK : RC_CCLIENT_FLEX_TRITS;
```
You can go further and check that `len == NUM_TRYTES_SERIALIZED_TRANSACTION `

thibault-martinez commented on cclient/request/broadcast_transactions.c:
```diff
@@ -22,14 +22,20 @@ void broadcast_transactions_req_free(broadcast_transactions_req_t** const req) {
+  if (len) {
+    hash8019_queue_push(&req->trytes, trits);
+  }
+  return len ? RC_OK : RC_CCLIENT_FLEX_TRITS;
```
Same check applies for this and all following

thibault-martinez commented on cclient/request/broadcast_transactions.h:
```diff
@@ -16,12 +16,12 @@ extern "C" {
 typedef struct broadcast_transactions_req_s {
   // List of raw transactions data to be broadcasted
-  flex_hash_array_t* trytes;
+  hash8019_queue_t trytes;
```
I think order does not matter for this call, you can use a stack for lower memory impact.

thibault-martinez commented on cclient/request/get_trytes.c:
```diff
@@ -20,17 +20,15 @@ void get_trytes_req_free(get_trytes_req_t** const req) {
-  req->hashes = flex_hash_array_append(req->hashes, hash);
-  return req;
+retcode_t get_trytes_req_add_hash(get_trytes_req_t* const req,
+                                  flex_trit_t const* const hash) {
```
Why are some `_*add` functions taking trytes and some flex_trits ?

oopsmonk replied:
the main idea is the request/response should take flex_trit_t, but `tyrtes` is provided for the convenience in test cases at json_serializer and ciri api. I think this function should be removed or changed it to take `tyrtes` for consistency.

thibault-martinez commented on cclient/request/store_transactions.h:
```diff
@@ -16,13 +16,13 @@ extern "C" {
 typedef struct store_transactions_req_s {
   // List of raw transactions data to be stored
-  flex_hash_array_t* trytes;
+  hash8019_queue_t trytes;
```
Order does not matter as well, maybe use a stack ?

oopsmonk replied:
sure, thanks :-)

Commit:
cclient: hash stack for broadcast_transactions, store_transactions.

Commit:
cclient: removing `tryte_t` operation in response/request.

Commit:
ciri: fixing test_get_tips failed.
